### PR TITLE
Adding preliminary url-parse logic (working!!)

### PR DESCRIPTION
**Pull request by sollipse**
Example of a working url: 
http://localhost:4200/#locations=JPN,ITA&num_cases_cutoff=10#locations=JPN,Alameda+County++C++USA&num_cases_cutoff=10

Some assumptions I've made: 
1. We want to keep the URL human-readble, with a minimum of url-encoding if possible. 
2. The HASH character (#) is not a naturally occurring part of a geographic location. 
3. Double spaces ("  ") can be substituted for ", "

Given our limited use case, I __think__ these are reasonable assumptions to be making, but I'm open to pivoting!

### CODE REVIEW

File: src/app/multi-line-chart/multi-line-chart.component.ts [hunk 1]
```diff
@@ -22,6 +22,7 @@ export class MultiLineChartComponent implements OnInit {
   // - all the serieses are of the same length & the same dates
   @Input() data: Series[];
   @Input() animate: boolean = false;
+  @Input() locations: string[];
```
Comment (prendradjaja) [inline src/app/multi-line-chart/multi-line-chart.component.ts]: Unused?

File: src/app/app.component.ts [hunk 2]
```diff
@@ -17,29 +18,29 @@ const MY_LOCATIONS = ['Alameda County, CA, USA', 'USA', 'ITA', 'JPN'];
...
+          if (!data[location]) continue;
+          graphData.push({
+            name: location,
+            values: data[location].map(item => item.cases),
+            comments: data[location].map(item => item.date.toDateString()),
+          });
+        }
+        this.graphs.push(graphData);
       }
+      console.log(this.graphs);
```
Comment (prendradjaja) [inline src/app/app.component.ts]: accidentally left in console.log?

File: src/app/url-parser.service.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+};
+
+@Injectable({
+  providedIn: 'root',
+})
+export class UrlParserService {
+  graphDefinitions: CovidGraphDefinition[];
+  constructor() {
+    let parsedQueryString = window.location.href
+      .slice(window.location.href.indexOf('#') + 1)
```
Comment (prendradjaja) [inline src/app/url-parser.service.ts]: Optional: I _think_ you can replace this whole expression with `window.location.hash`.

Alternatively: Or you can use Angular's tooling (ActivatedRoute.fragment). (Or we can merge this and I can show you that by putting up a followup PR!)

In any case heck ya this works yolo 🚢

**Comment by prendradjaja** [comment]
A couple minor things but this looks great! Nice. Looks like `qs` has no dependencies, which is great! Thanks for this :D

File: src/app/url-parser.service.ts (hunk 3 above)
Comment (sollipse) [reply src/app/url-parser.service.ts]: Was uncertain how location.hash would work with multiple hashes, but looks like it just returns the full string after the first encountered hash ... which is fine by me

File: src/app/app.component.ts (hunk 2 above)
Comment (sollipse) [reply src/app/app.component.ts]: yeppers

File: src/app/multi-line-chart/multi-line-chart.component.ts (hunk 1 above)
Comment (sollipse) [reply src/app/multi-line-chart/multi-line-chart.component.ts]: yep

File: src/app/app.component.html [hunk 4]
```diff
@@ -1,8 +1,4 @@
-<ng-container *ngIf="graphs == null; else dataAvailable">
-  <img
-    src="https://cdn.dribbble.com/users/1680472/screenshots/5697390/google-loader.gif"
-  />
-</ng-container>
+<ng-container *ngIf="graphs == null; else dataAvailable"> </ng-container>
```
Comment (prendradjaja) [inline src/app/app.component.html:1]:
```suggestion
```

File: src/app/url-parser.service.ts [hunk 5]
```diff
@@ -50,23 +51,33 @@ export type CovidGraphDefinition = {
...
+      ...userDefinedGraphProperties,
+      locations: replaceDashWithCommaSpace(
+        userDefinedGraphProperties.locations || []
+      ),
+    }));
+
 @Injectable({
   providedIn: 'root',
 })
 export class UrlParserService {
```
Comment (prendradjaja) [inline src/app/url-parser.service.ts:56]: Actually can you revert this last commit for now (and then you can just merge this PR straight away) and re-open as a separate PR with just this commit? I have some ideas of how to improve this a bit, but that'll be slightly more involved

File: src/app/app.component.html (hunk 4 above)
Comment (prendradjaja) [inline src/app/app.component.html]: (this change request can be left to the next PR)

----

Oo now that the ng-container is empty, let's change this like so:
```diff
-<ng-container *ngIf="condition; else other"> </ng-container>
-<ng-template #other> STUFF </ng-template>
+<ng-container *ngIf="!condition"> STUFF </ng-container>
```

(same as...)

```diff
-if (condition) {}
-else { STUFF }
+if (!condition) { STUFF }
```

**Pull request merged**